src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -807,113 +910,141 @@ def main():
-                    env,
-                    workspace_id,
-                    load_id,
-                    collections_source_dir,
-                    source_dir,
-                    upload_result
-                )
-                obj_name = upload_result.genome_tuple.obj_name
-                wait_to_upload_objs.pop(obj_name)
+def _fetch_objs_to_upload(
+        env: str,
+        ws: Workspace,
+        workspace_id: int,
+        load_id: str,
+        external_coll_src_dir: str,
+        ws_coll_src_dir: str,
+        source_dir: str,
+        upload_file_ext: list[str],
+        asu_client: AssemblyUtil,
+        job_data_dir: str
+):
+    """
+    Fetch objects to be uploaded to the workspace.
 
-            print("Recovery process completed ...")
+    Check if the objects are already uploaded to the workspace and perform recovery if needed.
+    """
+    count, wait_to_upload_objs = _fetch_objects_to_upload(
+        env, workspace_id, load_id, external_coll_src_dir, upload_file_ext)
 
-        if not wait_to_upload_objs:
-            print(
-                f"All {count} files already exist in workspace {workspace_id}"
-            )
+    # check if the objects are already uploaded to the workspace
+    obj_names_processed = _check_existing_uploads_and_recovery(
+        ws,
+        env,
+        workspace_id,
+        load_id,
+        ws_coll_src_dir,
+        source_dir,
+        wait_to_upload_objs,
+        asu_client,
+        job_data_dir,
+    )
+
+    for obj_name in obj_names_processed:
+        wait_to_upload_objs.pop(obj_name)
+
+    if not wait_to_upload_objs:
+        print(f"All {count} files already exist in workspace {workspace_id}")
+        return wait_to_upload_objs, count
+
+    wtus_len = len(wait_to_upload_objs)
+    print(f"Originally planned to upload {count} object files")
+    print(f"Detected {count - wtus_len} object files already exist in workspace")
+
+    return wait_to_upload_objs, count
+
+
+def main():
+
+    (workspace_id, kbase_collection, source_version, root_dir, token_filepath,
+     upload_file_ext, batch_size, cbs_max_tasks, au_service_ver, gfu_service_ver,
+     keep_job_dir, catalog_admin, load_id, env, kb_base_url) = _get_parser_args()
+
+    username = os.getlogin()
+    job_dir, external_coll_src_dir, ws_coll_src_dir, source_dir = _prepare_directories(
+        root_dir, username, kbase_collection, source_version, env, workspace_id
+    )
+
+    proc = None
+    conf = None
+    try:
+        proc, conf = _setup_and_start_services(job_dir, kb_base_url, token_filepath, cbs_max_tasks, catalog_admin)
+        if not proc or not conf:
+            print("Failed to start services. Exiting ...")
             return
 
-        wtus_len = len(wait_to_upload_objs)
-        print(f"Originally planned to upload {count} object files")
-        print(f"Detected {count - wtus_len} object files already exist in workspace")
+        ws, asu_client, gfu_client = _create_ws_clients(
```

Comment:
maybe rename to `_create_clients` if asu and gfu are not `ws` clients

Reply:
👍

Reply:
👍